### PR TITLE
s3select: not between operator

### PR DESCRIPTION
**Pull request by albin-antony**
Support not between operator
expression [NOT] between expression AND expression

Signed-off-by: Albin Antony <aantony@redhat.com>

### CODE REVIEW

File: include/s3select.h [hunk 1]
```diff
@@ -626,14 +626,16 @@ struct s3select : public bsc::grammar<s3select>
...
-      special_predicates = (is_null) | (is_not_null) | (between_predicate) | (in_predicate) | (like_predicate);
+      special_predicates = (is_null) | (is_not_null) | (between_predicate) | (not_between) | (in_predicate) | (like_predicate);
 
       is_null = ((factor) >> S3SELECT_KW("is") >> S3SELECT_KW("null"))[BOOST_BIND_ACTION(push_is_null_predicate)];
 
       is_not_null = ((factor) >> S3SELECT_KW("is") >> S3SELECT_KW("not") >> S3SELECT_KW("null"))[BOOST_BIND_ACTION(push_is_null_predicate)];
 
       between_predicate = (arithmetic_expression >> S3SELECT_KW("between") >> arithmetic_expression >> S3SELECT_KW("and") >> arithmetic_expression)[BOOST_BIND_ACTION(push_between_filter)];
 
+      not_between = (arithmetic_expression >> S3SELECT_KW("not") >> S3SELECT_KW("between") >> arithmetic_expression >> S3SELECT_KW("and") >> arithmetic_expression)[BOOST_BIND_ACTION(push_between_filter)];
```
Comment (galsalomon66) [inline include/s3select.h:637]: why not use a specific builder? (push_not_between_builde)

File: include/s3select_functions.h [hunk 2]
```diff
@@ -1436,6 +1438,25 @@ struct _fn_between : public base_function
...
 
+struct _fn_not_between : public base_function
+{
+
+  value res;
+  _fn_between between_op;
+  
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    between_op(args,result);
```
Comment (galsalomon66) [inline include/s3select_functions.h:1449]: :+1: 
re-use of `between_op`

File: test/s3select_test.cpp [hunk 3]
```diff
@@ -704,6 +704,23 @@ TEST(TestS3selectFunctions, between)
...
+  std::string input;
+  size_t size = 128;
+  generate_rand_columns_csv(input, size);
+  const std::string input_query_1 = "select count(0) from stdin where int(_1) not between int(_2) and int(_3);";
+
+  std::string s3select_result_1 = run_s3select(input_query_1,input);
+
+  const std::string input_query_2 = "select count(0) from stdin where int(_1) < int(_2) or int(_1) > int(_3);";
+
+  std::string s3select_result_2 = run_s3select(input_query_1,input);
```
Comment (galsalomon66) [inline test/s3select_test.cpp:719]: the queries seem to be correct (same semantics)
but
using the same query `input_query_1` for both runs.

File: include/s3select.h [hunk 4]
```diff
@@ -1201,6 +1203,19 @@ void push_between_filter::builder(s3select* self, const char* a, const char* b)
 
   std::string between_function("#between#");
 
+  bool between{true};
+
+  for(size_t i=0;i<token.size();i++)
+  {
+    bsc::parse_info<> info = bsc::parse(token.c_str()+i, ( bsc::str_p("not") >> bsc::str_p("between")) , bsc::space_p);
```
Comment (galsalomon66) [inline include/s3select.h:1210]: please validate that this part works as expected.
i.e. it accepts "not between" as it should

**Comment by galsalomon66** [comment]
please review the comments.
verify that all tests result correctly.

File: include/s3select.h (hunk 1 above)
Comment (albin-antony) [reply include/s3select.h:637]: There was a bug in the current implementation. Using a specific builder for not_between solves the issue. But, I am wondering Why not use the existing between builder similar to is null and is not null

File: test/s3select_test.cpp (hunk 3 above)
Comment (albin-antony) [reply test/s3select_test.cpp]: My bad! Updated the commit with the specific builder. Please take a look

File: include/s3select.h (hunk 1 above)
Comment (galsalomon66) [reply include/s3select.h]: > There was a bug in the current implementation. Using a specific builder for not_between solves the issue. But, I am wondering Why not use the existing between builder similar to is null and is not null

good question.
the scan-statement for is-not-null is different from the "not between" 
the "not between" is a complex expression compared to "is-not-null", the scan-statement for it should be different. 

on a side note its better to refactor the is-not-null builder, it's better not to use parser-within-parser

File: include/s3select.h [hunk 5]
```diff
@@ -1218,6 +1226,29 @@ void push_between_filter::builder(s3select* self, const char* a, const char* b)
   self->getAction()->exprQ.push_back(func);
 }
 
+void push_not_between_filter::builder(s3select* self, const char* a, const char* b) const
```
Comment (galsalomon66) [inline include/s3select.h:1229]: :+1:

File: test/s3select_test.cpp [hunk 6]
```diff
@@ -1375,6 +1392,14 @@ TEST(TestS3selectFunctions, test_date_time_expressions)
...
   ASSERT_NE(s3select_result_14, failure_sign);
   EXPECT_EQ(s3select_result_13, s3select_result_14);
+
+  std::string input_query_15 = "select to_string(to_timestamp(_1), 'y,M,H,m') from stdin where cast(to_string(to_timestamp(_1), 'd') as int) < 1 or cast(to_string(to_timestamp(_1), 'd') as int) > 10;";
+  std::string s3select_result_15 = run_s3select(input_query_15, input);
+  ASSERT_NE(s3select_result_15, failure_sign);
+  std::string input_query_16 = "select extract(year from to_timestamp(_1)), extract(month from to_timestamp(_1)), extract(hour from to_timestamp(_1)), extract(minute from to_timestamp(_1)) from stdin where  int(substring(_1, 9, 2)) not between 1 and 10;";
+  std::string s3select_result_16 = run_s3select(input_query_16, input);
+  ASSERT_NE(s3select_result_16, failure_sign);
+  EXPECT_EQ(s3select_result_15, s3select_result_16);
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1402]: :+1: 
good test coverage, the query syntax is quite different

File: test/s3select_test.cpp [hunk 7]
```diff
@@ -1795,6 +1820,18 @@ TEST(TestS3selectFunctions, tuefalse_like_expressions)
...
+
+  std::string s3select_result_9 = run_s3select(input_query_9,input);
+
+  ASSERT_NE(s3select_result_9,failure_sign);
+
+  const std::string input_query_10 = "select count(*) from stdin where (substring(_1,char_length(_1),1) not between \"r\" and \"s\") = true or (substring(_1,char_length(_1)-1,1) = \"r\") = false;";
+
+  std::string s3select_result_10 = run_s3select(input_query_10,input);
+
+  ASSERT_EQ(s3select_result_9, s3select_result_10);
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1834]: please check this test, the results are empty.

**Comment by galsalomon66** [comment]
please review the comments.
upon creating a new test(especially with `count`), it's a must to check results, it should not be empty.

File: test/s3select_test.cpp (hunk 7 above)
Comment (albin-antony) [reply test/s3select_test.cpp:1834]: Strange, I am getting 10k for both. I guess the current data set used for these queries doesn't contain those patterns

Comment (galsalomon66) [reply test/s3select_test.cpp:1834]: the query should not return also the full data set (the query is not distinctive enough).

it's quite easy to verify results.

File: include/s3select.h [hunk 8]
```diff
@@ -1218,6 +1226,29 @@ void push_between_filter::builder(s3select* self, const char* a, const char* b)
...
+{
+  std::string token(a, b);
```
Comment (mattbenjamin) [inline include/s3select.h:1231]: looks like token is unused;  I also don't recognize this std::string constructor.  What are we trying to do here?

File: include/s3select.h [hunk 9]
```diff
@@ -1218,6 +1226,29 @@ void push_between_filter::builder(s3select* self, const char* a, const char* b)
...
+
+  std::string not_between_function("#not_between#");
```
Comment (mattbenjamin) [inline include/s3select.h:1233]: is "not_between_function" a kind of string constant? if so, it is conventional (and more efficient) to declare it as an actual string (or string_view) constant

File: include/s3select.h (hunk 8 above)
Comment (albin-antony) [reply include/s3select.h]: Removed it. It was part of the previous commit

File: include/s3select.h (hunk 9 above)
Comment (galsalomon66) [reply include/s3select.h:1231]: yes, it could/should be constant.

this method is been called upon parsing only, a few times.

Comment (mattbenjamin) [reply include/s3select.h:1231]: static constexpr const char* not_between = "#not_between#";

File: test/s3select_test.cpp (hunk 7 above)
Comment (galsalomon66) [reply test/s3select_test.cpp:1835]: ok, it returns results 
what was the change?

File: test/s3select_test.cpp [hunk 10]
```diff
@@ -594,8 +594,9 @@ void generate_csv_trim(std::string& out, size_t size) {
 void generate_csv_like(std::string& out, size_t size) {
   // schema is: int, float, string, string
   std::stringstream ss;
+  auto r = [](){ int x=rand()%1000;if (x<500) return std::string("hai"); else return std::string("fooaeioubrs");};
   for (auto i = 0U; i < size; ++i) {
-    ss << "fooaeioubrs" << "," << std::endl;
+    ss << r() << "," << std::endl;
```
Comment (galsalomon66) [inline test/s3select_test.cpp:599]: :+1:

File: test/s3select_test.cpp (hunk 7 above)
Comment (galsalomon66) [reply test/s3select_test.cpp:1835]: ok, got that.

**Review by galsalomon66 — COMMENTED** [review]
looks good, and also the test coverage.

**Comment by galsalomon66** [comment]
please fix comments for constant and un-used variables

File: include/s3select.h [hunk 11]
```diff
@@ -1218,6 +1226,27 @@ void push_between_filter::builder(s3select* self, const char* a, const char* b)
   self->getAction()->exprQ.push_back(func);
 }
 
+void push_not_between_filter::builder(s3select* self, const char* a, const char* b) const
+{
+  const std::string not_between_function("#not_between#");
```
Comment (galsalomon66) [inline include/s3select.h:1231]: please use `static constexpr const std::string_view not_between_function("#not_between#");`
need to replace `c_str()` with `data()`

File: include/s3select.h [hunk 12]
```diff
@@ -15,6 +15,7 @@
 #include <boost/function.hpp>
 #include <boost/bind.hpp>
 #include <functional>
+#include <string_view>
```
Comment (galsalomon66) [inline include/s3select.h:18]: is it a must? 
s3select already using `string_view`

**Review by galsalomon66 — APPROVED** [review]

File: include/s3select.h (hunk 12 above)
Comment (mattbenjamin) [reply include/s3select.h]: if we are using string_view in the header itself, then yes, <string_view> should be included here for clarity (include what you use)

**Pull request merged**